src/otp-input/otp-input.tsx

```diff
@@ -0,0 +1,212 @@
+    // CONST, STATE, REF
+    // =============================================================================
+    const {
+        disabled,
+        onClick,
+        styleType = "secondary",
+        ...otherCtaProps
+    } = actionButtonProps ?? {};
+    const inputRefs = useRef<(HTMLInputElement | null)[]>([]);
+
+    const [otpValues, setOtpValues] = useState<string[]>(
+        new Array(numOfInput).fill("")
+    );
+    const [countDown, setCountDown] = useState<number>(cooldownDuration);
+    const [lastCtaTimestamp, setLastCtaTimestamp] = useState<Date | undefined>(
+        new Date()
+    );
+
+    // =============================================================================
+    // EFFECTS
+    // =============================================================================
+    useEffect(() => {
+        inputRefs?.current[0]?.focus();
+        document.addEventListener("paste", handlePaste);
+        return () => document.removeEventListener("paste", handlePaste);
+    }, []);
+
+    useEffect(() => {
+        if (countDown === cooldownDuration) {
+            if (onCooldownStart) onCooldownStart();
+            beginCountdown();
+        }
+    }, [countDown]);
+
+    // =============================================================================
+    // EVENT HANDLERS
+    // =============================================================================
+
+    const handleChange =
+        (index: number) => (event: React.ChangeEvent<HTMLInputElement>) => {
+            const value = event.target.value.replace(/[^0-9]/g, "");
+            if (validateUserInput(value)) {
+                const newOtpValues = [...otpValues];
+
+                newOtpValues[index] = value.substring(value.length - 1);
+                inputRefs.current[index + 1]?.focus();
+
+                setOtpValues(newOtpValues);
+
+                if (onChange) {
+                    onChange(newOtpValues.join(""));
+                }
+            }
+        };
+
+    const handleKeyDown =
+        (index: number) => (event: React.KeyboardEvent<HTMLInputElement>) => {
+            const { key, code } = event;
+
+            const newOtpValues = [...otpValues];
+            if (
+                (code === "Backspace" || key === "Backspace") &&
+                newOtpValues[index] !== ""
+            ) {
+                newOtpValues[index] = "";
+            } else if (code === "Backspace" || key === "Backspace") {
+                newOtpValues[index - 1] = "";
+                inputRefs.current[index - 1]?.focus();
+            }
+
+            setOtpValues(newOtpValues);
+
+            if (onChange) {
+                onChange(newOtpValues.join(""));
+            }
+        };
+
+    const handlePaste = (event: ClipboardEvent): void => {
+        const pastedValue = event.clipboardData.getData("text");
+        if (pastedValue && validateUserInput(pastedValue, numOfInput)) {
+            setOtpValues(pastedValue.split(""));
+            if (onChange) {
```

Comment:
risk of stale props: `numOfInput` (unlikely) or `onChange` (possibly)

if this is an issue, these variables will need to be ref'd to keep up to date